open_spiel/games/nim_test.cc

```diff
@@ -0,0 +1,145 @@
+// Copyright 2019 DeepMind Technologies Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "open_spiel/spiel.h"
+#include "open_spiel/algorithms/value_iteration.h"
+#include "open_spiel/tests/basic_tests.h"
+
+namespace open_spiel {
+namespace nim {
+namespace {
+
+namespace testing = open_spiel::testing;
+namespace algorithms = open_spiel::algorithms;
+
+void BasicNimTests() {
+  testing::LoadGameTest("nim");
+  testing::RandomSimTest(*LoadGame("nim"), 100);
+  testing::RandomSimTest(
+      *LoadGame("nim",
+                {
+                    {"pile_sizes", GameParameter("100;200;300")},
+                }),
+      10);
+  testing::RandomSimTest(
+      *LoadGame("nim",
+                {
+                    {"pile_sizes", GameParameter("10000;2000;3000;12414;1515;53252;1;35126")},
+                }),
+      10);
+  testing::RandomSimTest(
+      *LoadGame("nim",
+                {
+                    {"pile_sizes", GameParameter("1;2;3;4;5;6;7;8;9;10")},
+                    {"is_misere", GameParameter(false)},
+                }),
+      10);
+}
+
+void SinglePileNormalTest() {
+  std::shared_ptr<const Game> game = LoadGame(
+      "nim", {
+          {"pile_sizes", GameParameter("100")},
+          {"is_misere", GameParameter(false)},
+      });
+  std::unique_ptr<State> state = game->NewInitialState();
+  std::vector<Action> actions = state->LegalActions();
+  SPIEL_CHECK_EQ(actions.size(), 100);
+
+  state->ApplyAction(actions.back());
+  SPIEL_CHECK_EQ(state->IsTerminal(), 1);
+  SPIEL_CHECK_EQ(state->PlayerReturn(0), 1);
+  SPIEL_CHECK_EQ(state->PlayerReturn(1), -1);
+}
+
+void SinglePileMisereTest() {
+  std::shared_ptr<const Game> game = LoadGame(
+      "nim", {
+          {"pile_sizes", GameParameter("100")},
+      });
+  std::unique_ptr<State> state = game->NewInitialState();
+  std::vector<Action> actions = state->LegalActions();
+  SPIEL_CHECK_EQ(actions.size(), 100);
+
+  state->ApplyAction(actions.back());
+  SPIEL_CHECK_EQ(state->IsTerminal(), 1);
+  SPIEL_CHECK_EQ(state->PlayerReturn(0), -1);
+  SPIEL_CHECK_EQ(state->PlayerReturn(1), 1);
+}
+
+void VISinglePileMisereTest() {
```

Comment:
As I understand it, the same test but only with ```is_misere = false``` should yield ```values["100"]``` being equal to 1 as well as with ```is_misere = true``` (since an agent can take everything from the only pile)

However, the ```ValueIteration``` returns -1 for the case of ```is_misere = false```. I manually ran an mcts example as well as did the test ```SinglePileNormalTest()``` above, and both of them seem to work fine, so I'm not sure why it's happening. @lanctot Could you take a look please?

Reply:
Yes I believe it is because NimState::ToString is missing the current player, so the states are being aliased across players. 

Seems to be included in the Observation but missing from ToString (and ObservationString by extension).

Reply:
Note: you will need to update the playthrough as well.

Reply:
thanks, it seemed to work!